pkg/api/handlers/utils/containers.go

```diff
@@ -27,32 +26,92 @@ func WaitContainer(w http.ResponseWriter, r *http.Request) (int32, error) {
 		Error(w, "Something went wrong.", http.StatusBadRequest, errors.Wrapf(err, "failed to parse parameters for %s", r.URL.String()))
 		return 0, err
 	}
-	if _, found := r.URL.Query()["interval"]; found {
-		interval, err = time.ParseDuration(query.Interval)
-		if err != nil {
-			InternalServerError(w, err)
-			return 0, err
-		}
-	} else {
-		interval, err = time.ParseDuration("250ms")
-		if err != nil {
-			InternalServerError(w, err)
-			return 0, err
-		}
-	}
-	condition := define.ContainerStateStopped
+
+	var condition string
 	if _, found := r.URL.Query()["condition"]; found {
-		condition, err = define.StringToContainerStatus(query.Condition)
-		if err != nil {
-			InternalServerError(w, err)
+		condition = query.Condition
+		if condition != "next-exit" && condition != "removed" && condition != "not-running" && condition != "" {
+			err := errors.New("invalid value for the 'condition' parameter, valid values are: 'not-running', 'next-exit', 'removed'")
+			BadRequest(w, "condition", condition, err)
 			return 0, err
 		}
 	}
+
 	name := GetName(r)
 	con, err := runtime.LookupContainer(name)
 	if err != nil {
 		ContainerNotFound(w, name, err)
 		return 0, err
 	}
-	return con.WaitForConditionWithInterval(interval, condition)
+
+	beforeWait()
+
+	ctx := r.Context()
+	var code int32
+	switch condition {
+	case "next-exit":
+		code, err = waitNextExit(ctx, con)
+	case "removed":
+		code, err = waitRemoved(ctx, con)
+	default:
+		code, err = waitNotRunning(ctx, con)
+	}
+
+	if err != nil {
+		InternalServerError(w, err)
+		return 0, err
+	}
+
+	return code, nil
+}
+
+func waitForOneOf(ctx context.Context, con *libpod.Container, stats ...define.ContainerStatus) (int32, error) {
+	newCtx, cancelFn := context.WithCancel(ctx)
+	ch := make(chan struct {
+		code int32
+		err  error
+	}, 1)
+
+	for _, s := range stats {
+		go func(s define.ContainerStatus) {
+			code, err := con.WaitForConditionWithInterval(newCtx, interval, s)
```

Comment:
Probably better to rewrite `WaitForConditionWithInterval` to accept multiple conditions - Libpod API is allowed to change, and it seems a lot more logical to handle this there instead of spawning numerous goroutines to handle it.

Reply:
@mheon given the problem that arisen I am considering reverting the change that made `WaitForConditionWithInterval()` variadic.